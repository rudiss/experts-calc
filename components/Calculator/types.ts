import { Dispatch, SetStateAction } from "react";

export enum BTN_ACTIONS {
  ADD = "ADD",
  THEME = "THEME",
  CALC = "CALC",
  DELETE = "DELETE",
}

export type buttonOptionType = {
  display: string | number;
  action: BTN_ACTIONS;
  class: string;
};

export interface IOperationActionProps {
  item: buttonOptionType;
  expression: string;
  setExpression: Dispatch<SetStateAction<string>>;
  expElement: HTMLDivElement;
  resultElement: HTMLDivElement;
  setDarkTheme: Dispatch<SetStateAction<boolean>>;
}
