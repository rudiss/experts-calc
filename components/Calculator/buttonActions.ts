type BTN_ACTIONS = {
    ADD: "ADD" | string,
    THEME: "THEME" | string,
    CALC: "CALC" | string,
    DELETE: "DELETE" | string,
  };

  export const BTN_ACTIONS: BTN_ACTIONS = {
    ADD: "ADD",
    THEME: "THEME",
    CALC: "CALC",
    DELETE: "DELETE",
  };

  export type buttonOptionType = {
    display: string | number,
    action: typeof BTN_ACTIONS | string,
    class: string,
  };

  export type buttonsOptionsTypes = buttonOptionType[];

  export const buttonsOptions: buttonsOptionsTypes = [
    {
      display: "C",
      action: BTN_ACTIONS.DELETE,
      class: "btn-operator",
    },
    {
      display: "(",
      action: BTN_ACTIONS.ADD,
      class: "btn-operator",
    },
    {
      display: ")",
      action: BTN_ACTIONS.ADD,
      class: "btn-operator",
    },
    {
      display: "/",
      action: BTN_ACTIONS.ADD,
      class: "btn-operator",
    },
    {
      display: "7",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "8",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "9",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "x",
      action: BTN_ACTIONS.ADD,
      class: "btn-operator",
    },
    {
      display: "4",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "5",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "6",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "-",
      action: BTN_ACTIONS.ADD,
      class: "btn-operator",
    },
    {
      display: "1",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "2",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "3",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "+",
      action: BTN_ACTIONS.ADD,
      class: "btn-operator",
    },
    {
      display: "sun.svg",
      action: BTN_ACTIONS.THEME,
      class: "",
    },
    {
      display: "0",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: ".",
      action: BTN_ACTIONS.ADD,
      class: "",
    },
    {
      display: "=",
      action: BTN_ACTIONS.CALC,
      class: "btn-operator",
    },
  ];
