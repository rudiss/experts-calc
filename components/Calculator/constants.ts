export const CALCULATOR_EXPRESSION_TEST_ID = "calculatorExpression";
export const CALCULATOR_RESULT_TEST_ID = "calculatorResult";

export const CALCULATOR_BUTTONS_TEST_ID = "calculatorButtons";
