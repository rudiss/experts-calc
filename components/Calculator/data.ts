import { BTN_ACTIONS, buttonOptionType } from "./types";

export const buttonsOptions: buttonOptionType[] = [
  {
    display: "C",
    action: BTN_ACTIONS.DELETE,
    class: "btn-operator",
  },
  {
    display: "(",
    action: BTN_ACTIONS.ADD,
    class: "btn-operator",
  },
  {
    display: ")",
    action: BTN_ACTIONS.ADD,
    class: "btn-operator",
  },
  {
    display: "/",
    action: BTN_ACTIONS.ADD,
    class: "btn-operator",
  },
  {
    display: "7",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "8",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "9",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "x",
    action: BTN_ACTIONS.ADD,
    class: "btn-operator",
  },
  {
    display: "4",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "5",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "6",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "-",
    action: BTN_ACTIONS.ADD,
    class: "btn-operator",
  },
  {
    display: "1",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "2",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "3",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "+",
    action: BTN_ACTIONS.ADD,
    class: "btn-operator",
  },
  {
    display: "sun.svg",
    action: BTN_ACTIONS.THEME,
    class: "",
  },
  {
    display: "0",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: ".",
    action: BTN_ACTIONS.ADD,
    class: "",
  },
  {
    display: "=",
    action: BTN_ACTIONS.CALC,
    class: "btn-operator",
  },
];
