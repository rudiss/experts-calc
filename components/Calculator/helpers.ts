import { IOperationActionProps, BTN_ACTIONS } from "./types";

const handleAddAnimationSpan = (content: string, expression: any) => {
  const span = document.createElement("span");

  span.innerHTML = content;
  span.style.opacity = "0";
  expression?.appendChild(span);

  const width = span.offsetWidth + "px";
  span.style.width = "0";

  setTimeout(() => {
    span.style.opacity = "1";
    span.style.width = width;
  }, 100);
};

const setActionAdd = ({
  item,
  expression,
  expElement,
  setExpression,
}: IOperationActionProps) => {
  handleAddAnimationSpan(String(item.display), expElement);

  console.log(expression);
  const operator = item.display !== "x" ? item.display : "*";
  setExpression(expression + operator);
};

const setActionDelete = ({
  resultElement,
  expElement,
  setExpression,
}: IOperationActionProps) => {
  resultElement.innerHTML = "";
  expElement.innerHTML = "";

  setExpression("");
};

const setActionCalc = ({
  setExpression,
  expression,
  expElement,
  resultElement,
}: IOperationActionProps) => {
  if (expression.trim().length <= 0 || !expElement.parentNode) return;

  resultElement.remove();

  const cloneNode = expElement.cloneNode(true) as HTMLDivElement;
  expElement.parentNode.appendChild(cloneNode);

  const transform = `translateY(${
    -(expElement?.offsetHeight + 10) + "px"
  }) scale(0.4)`;

  try {
    let res = eval(expression);

    setExpression(String(res));
    setTimeout(() => {
      cloneNode.style.transform = transform;
      expElement.innerHTML = "";

      const calcAnimationSpan = String(Math.floor(res * 100000000) / 100000000);
      handleAddAnimationSpan(calcAnimationSpan, expElement);
    }, 200);
  } catch {
    setTimeout(() => {
      cloneNode.style.transform = transform;
      cloneNode.innerHTML = "Not a number";
    }, 200);
  } finally {
    console.log("calc complete");
  }
};

const setActionTheme = ({ setDarkTheme }: IOperationActionProps) => {
  document.body.classList.toggle("dark");
  setDarkTheme((prevValue) => !prevValue);
};

export const selectActionItemByInputClick = (props: IOperationActionProps) =>
  ({
    [BTN_ACTIONS.ADD]: () => setActionAdd(props),
    [BTN_ACTIONS.DELETE]: () => setActionDelete(props),
    [BTN_ACTIONS.CALC]: () => setActionCalc(props),
    [BTN_ACTIONS.THEME]: () => setActionTheme(props),
  }[props.item.action]());
