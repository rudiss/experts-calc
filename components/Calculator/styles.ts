import styled from "styled-components";

export const Container = styled.div`
  padding: 20px;
  height: 700px;
  width: 400px;
  overflow: hidden;
  border-radius: 20px;
  background-color: var(--calc-bg);
  box-shadow: rgba(100, 100, 111, 0.2) 0px 7px 29px 0px;
  border: 1px solid var(--txt-white);

  display: flex;
  flex-direction: column;
  justify-content: flex-end;

  transition: all 0.8s ease;

  @media only screen and (max-width: 800px) {
    width: 100%;
    height: 100vh;
    border-radius: 0;
  }
`;
export const Result = styled.div`
  position: relative;
  margin-bottom: 20px;
  text-align: right;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: flex-end;
`;
export const Exp = styled.div`
  font-size: 3.5rem;
  line-height: 1;
  transform-origin: right bottom;
  transition: transform 0.3s ease;
  color: var(--calc-res-color);

  &:last-child {
    position: absolute;
    right: 0;
  }
  span {
    display: inline-block;
    overflow: hidden;
    transition: width 0.3s ease;
  }
`;

export const ButtonsContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  gap: 10px;
  opacity: 0;

  button {
    border: 0;
    outline: 0;
    border-radius: 10px;
    font-size: 1.25rem;
    font-weight: 500;
    background-color: transparent;
    color: var(--txt-color);

    &:last-child {
      color: var(--txt-white);
      background-image: linear-gradient(312deg, #cd6e36, #d400ff);
    }
  }
  .btn-operator {
    color: var(--calc-btn-color);
  }
`;
