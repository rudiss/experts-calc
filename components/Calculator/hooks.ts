import { selectActionItemByInputClick } from "./helpers";
import { useRef, useState } from "react";
import { buttonOptionType } from "./types";

export const useCalculator = () => {
  const expressionRef = useRef<HTMLInputElement>(null);
  const [expression, setExpression] = useState<string>("");
  const [darkTheme, setDarkTheme] = useState<boolean>(false);

  const handleInputAction = (item: buttonOptionType) => {
    const expElement = expressionRef.current as HTMLDivElement;
    const resultElement =
      expElement.parentNode &&
      (expElement.parentNode.querySelector("div:last-child") as HTMLDivElement);

    if (!expElement || !resultElement) return;

    selectActionItemByInputClick({
      item,
      expElement,
      resultElement,
      setExpression,
      expression,
      setDarkTheme,
    });
  };

  return {
    darkTheme,
    expressionRef,
    handleInputAction,
  };
};
