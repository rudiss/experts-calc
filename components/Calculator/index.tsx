import Image from "next/image";
import * as React from "react";
import { buttonOptionType } from "./types";
import { buttonsOptions } from "./data";
import { useCalculator } from "./hooks";

import { Container, Result, Exp, ButtonsContainer } from "./styles";
import {
  CALCULATOR_BUTTONS_TEST_ID,
  CALCULATOR_EXPRESSION_TEST_ID,
  CALCULATOR_RESULT_TEST_ID,
} from "./constants";

const { useRef, useEffect, useCallback } = React;

const Calculator: React.FC = () => {
  const { expressionRef, handleInputAction, darkTheme, expression } =
    useCalculator();

  const buttonsRef = useRef<HTMLInputElement>(null);

  const handleCalculatorButtonsStyle = useCallback(
    () =>
      setTimeout(() => {
        if (!buttonsRef.current) return;

        const calculatorButtons: HTMLButtonElement[] = Array.from(
          buttonsRef.current.querySelectorAll("button")
        );

        calculatorButtons.forEach(
          (element) => (element.style.height = element.offsetWidth + "px")
        );

        buttonsRef.current.style.opacity = "1";
      }, 250),
    [buttonsRef]
  );

  useEffect(() => {
    handleCalculatorButtonsStyle();
  }, [handleCalculatorButtonsStyle]);

  const handleCLickItem = useCallback(
    (item: buttonOptionType) => {
      handleInputAction(item);
    },
    [handleInputAction]
  );

  return (
    <Container>
      <Result className="calculator__result">
        <Exp
          ref={expressionRef}
          data-testid={CALCULATOR_RESULT_TEST_ID}
          role="input"
          className="calculator__result__exp"
          aria-label="Result"
          data-value={expression}
        />

        <Exp
          className="calculator__result__exp"
          data-testid={CALCULATOR_EXPRESSION_TEST_ID}
        />
      </Result>
      <ButtonsContainer
        ref={buttonsRef}
        data-testid={CALCULATOR_BUTTONS_TEST_ID}
        className="calculator__btns"
      >
        {buttonsOptions.map((button, index) => (
          <button
            key={index}
            className={button.class}
            onClick={() => handleCLickItem(button)}
          >
            {button.display !== "sun.svg" && button.display}
            {button.display === "sun.svg" && (
              <Image
                src={`/${!darkTheme ? "moon.png" : "sun.png"}`}
                alt={button.display}
                width={16}
                height={16}
              />
            )}
          </button>
        ))}
      </ButtonsContainer>
    </Container>
  );
};

export default Calculator;
