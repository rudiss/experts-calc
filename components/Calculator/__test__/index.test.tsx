import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import Calculator from "..";
import { CALCULATOR_RESULT_TEST_ID } from "../constants";

describe("<Calculator>", () => {
  it("should render the calculator", () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });

  it("should render the calculator and sum value", () => {
    render(<Calculator />);

    const buttonOne = screen.getByText(/1/);
    fireEvent.click(buttonOne);

    const buttonZero = screen.getByText(/0/g);
    fireEvent.click(buttonZero);

    const result = screen.getByTestId(CALCULATOR_RESULT_TEST_ID).textContent;

    expect(result).toEqual("10");
  });


  it("should render the calculator and clear value", () => {
    render(<Calculator />);

    const buttonOne = screen.getByText(/1/);
    fireEvent.click(buttonOne);

    const buttonZero = screen.getByText(/0/g);
    fireEvent.click(buttonZero);


    const buttonClean = screen.getByText(/C/g);
    fireEvent.click(buttonClean);

    const result = screen.getByTestId(CALCULATOR_RESULT_TEST_ID).textContent;

    expect(result).toEqual("");
  });
});
