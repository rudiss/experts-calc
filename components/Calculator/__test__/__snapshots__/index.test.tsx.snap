// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator> should render the calculator 1`] = `
<div>
  <div
    class="sc-b5fa8bf3-0 fKQtld"
  >
    <div
      class="sc-b5fa8bf3-1 YpViN calculator__result"
    >
      <div
        aria-label="Result"
        class="sc-b5fa8bf3-2 jPTnNe calculator__result__exp"
        data-testid="calculatorResult"
        data-value=""
        role="input"
      />
      <div
        class="sc-b5fa8bf3-2 jPTnNe calculator__result__exp"
        data-testid="calculatorExpression"
      />
    </div>
    <div
      class="sc-b5fa8bf3-3 uDyhv calculator__btns"
      data-testid="calculatorButtons"
    >
      <button
        class="btn-operator"
      >
        C
      </button>
      <button
        class="btn-operator"
      >
        (
      </button>
      <button
        class="btn-operator"
      >
        )
      </button>
      <button
        class="btn-operator"
      >
        /
      </button>
      <button
        class=""
      >
        7
      </button>
      <button
        class=""
      >
        8
      </button>
      <button
        class=""
      >
        9
      </button>
      <button
        class="btn-operator"
      >
        x
      </button>
      <button
        class=""
      >
        4
      </button>
      <button
        class=""
      >
        5
      </button>
      <button
        class=""
      >
        6
      </button>
      <button
        class="btn-operator"
      >
        -
      </button>
      <button
        class=""
      >
        1
      </button>
      <button
        class=""
      >
        2
      </button>
      <button
        class=""
      >
        3
      </button>
      <button
        class="btn-operator"
      >
        +
      </button>
      <button
        class=""
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2716%27%20height=%2716%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="sun.svg"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </button>
      <button
        class=""
      >
        0
      </button>
      <button
        class=""
      >
        .
      </button>
      <button
        class="btn-operator"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
